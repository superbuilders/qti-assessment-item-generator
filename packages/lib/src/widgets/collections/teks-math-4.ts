import type { WidgetCollection } from "@/widgets/collections/types"
import { allWidgetDefinitions } from "@/widgets/definitions"

export const teksMath4Collection: WidgetCollection<
	Pick<
		typeof allWidgetDefinitions,
		| "protractorAngleDiagram"
		| "nPolygon"
		| "dotPlot"
		| "pictograph"
		| "tapeDiagram"
		| "numberLineWithAction"
		| "fractionModelDiagram"
		| "quantityFractionalDiagram"
		| "symmetryDiagram"
		| "stemLeafPlot"
		| "dataTable"
		| "geometricPrimitiveDiagram"
		| "decoratedNumber"
	>,
	readonly [
		"protractorAngleDiagram",
		"nPolygon",
		"dotPlot",
		"pictograph",
		"tapeDiagram",
		"numberLineWithAction",
		"fractionModelDiagram",
		"quantityFractionalDiagram",
		"symmetryDiagram",
		"stemLeafPlot",
		"dataTable",
		"geometricPrimitiveDiagram",
		"decoratedNumber"
	]
> = {
	name: "teks-math-4",
	widgets: {
		protractorAngleDiagram: allWidgetDefinitions.protractorAngleDiagram,
		nPolygon: allWidgetDefinitions.nPolygon,
		dotPlot: allWidgetDefinitions.dotPlot,
		pictograph: allWidgetDefinitions.pictograph,
		tapeDiagram: allWidgetDefinitions.tapeDiagram,
		numberLineWithAction: allWidgetDefinitions.numberLineWithAction,
		fractionModelDiagram: allWidgetDefinitions.fractionModelDiagram,
		quantityFractionalDiagram: allWidgetDefinitions.quantityFractionalDiagram,
		symmetryDiagram: allWidgetDefinitions.symmetryDiagram,
		stemLeafPlot: allWidgetDefinitions.stemLeafPlot,
		dataTable: allWidgetDefinitions.dataTable,
		geometricPrimitiveDiagram: allWidgetDefinitions.geometricPrimitiveDiagram,
		decoratedNumber: allWidgetDefinitions.decoratedNumber
	},
	widgetTypeKeys: [
		"protractorAngleDiagram",
		"nPolygon",
		"dotPlot",
		"pictograph",
		"tapeDiagram",
		"numberLineWithAction",
		"fractionModelDiagram",
		"quantityFractionalDiagram",
		"symmetryDiagram",
		"stemLeafPlot",
		"dataTable",
		"geometricPrimitiveDiagram",
		"decoratedNumber"
	]
} as const
