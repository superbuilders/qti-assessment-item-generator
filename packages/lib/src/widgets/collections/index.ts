import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
