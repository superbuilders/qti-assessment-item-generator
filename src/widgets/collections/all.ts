import type { WidgetCollection } from "@/widgets/collections/types"
import { allWidgetDefinitions } from "@/widgets/definitions"

function toReadonlyKeys<T extends Record<string, unknown>>(
	obj: T
): ReadonlyArray<keyof T & string> {
	const keys = Object.keys(obj)
	// Narrow via a runtime check that keys are strings
	return keys.filter((k): k is keyof T & string => typeof k === "string")
}

export const allWidgetsCollection: WidgetCollection<
	typeof allWidgetDefinitions,
	ReadonlyArray<keyof typeof allWidgetDefinitions & string>
> = {
	name: "all",
	widgets: allWidgetDefinitions,
	widgetTypeKeys: toReadonlyKeys(allWidgetDefinitions)
} as const
