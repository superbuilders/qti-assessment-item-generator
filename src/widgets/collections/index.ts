import { allWidgetsCollection } from "./all"
import { fourthGradeMathCollection } from "./fourth-grade-math"
import { mathCoreCollection } from "./math-core"
import { scienceCollection } from "./science"
import { simpleVisualCollection } from "./simple-visual"
import { teksMath4Collection } from "./teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "./types"
