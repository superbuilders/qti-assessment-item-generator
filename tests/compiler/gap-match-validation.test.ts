import { describe, expect, test } from "bun:test"
import { compile } from "@/compiler/compiler"
import type { AssessmentItemInput } from "@/core/item"
import {
	MINIMAL_CORRECT_FEEDBACK,
	MINIMAL_INCORRECT_FEEDBACK
} from "@/testing/helpers/feedback-fixtures"
import { allWidgetsCollection } from "@/widgets/collections/all"

describe("Compiler: gapMatchInteraction validation", () => {
	const baseValidItem: AssessmentItemInput<[]> = {
		identifier: "gap-match-validation-item",
		title: "Gap Match Validation",
		responseDeclarations: [
			{
				identifier: "RESPONSE",
				cardinality: "multiple",
				baseType: "directedPair",
				correct: [{ source: "TOKEN_A", target: "GAP_1" }],
				allowEmpty: false
			}
		],
		body: [{ type: "interactionRef", interactionId: "gap_match" }],
		widgets: null,
		interactions: {
			gap_match: {
				type: "gapMatchInteraction",
				responseIdentifier: "RESPONSE",
				shuffle: true,
				content: [
					{
						type: "paragraph",
						content: [
							{ type: "text", content: "Fill blank: " },
							{ type: "gap", gapId: "GAP_1" }
						]
					}
				],
				gapTexts: [
					{
						identifier: "TOKEN_A",
						matchMax: 1,
						content: [{ type: "text", content: "A" }]
					}
				],
				gaps: [{ identifier: "GAP_1", required: true }]
			}
		},
		feedbackPlan: {
			mode: "combo",
			dimensions: [
				{
					responseIdentifier: "RESPONSE",
					kind: "binary"
				}
			],
			combinations: [
				{
					id: "FB__RESPONSE_CORRECT",
					path: [{ responseIdentifier: "RESPONSE", key: "CORRECT" }]
				},
				{
					id: "FB__RESPONSE_INCORRECT",
					path: [{ responseIdentifier: "RESPONSE", key: "INCORRECT" }]
				}
			]
		},
		feedback: {
			FEEDBACK__OVERALL: {
				RESPONSE: {
					CORRECT: { content: MINIMAL_CORRECT_FEEDBACK },
					INCORRECT: { content: MINIMAL_INCORRECT_FEEDBACK }
				}
			}
		}
	}

	test("should reject gapMatchInteraction when content is null", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.interactions.gap_match.content = null
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject gapMatchInteraction when content is an empty array", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.interactions.gap_match.content = []
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject when gap count in content mismatches gaps array (too few)", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.interactions.gap_match.gaps.push({
			identifier: "GAP_2",
			required: true
		}) // Declared 2, used 1
		invalidItem.responseDeclarations[0].correct.push({
			source: "TOKEN_A",
			target: "GAP_2"
		})
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject when gap count in content mismatches gaps array (too many)", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.interactions.gap_match.content[0].content.push({
			type: "gap",
			gapId: "GAP_1"
		}) // Used 2, declared 1
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject when a gapId in content is not declared in gaps array", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.interactions.gap_match.content[0].content[1].gapId =
			"GAP_UNDECLARED"
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject when a correct.target is not in gaps array", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.responseDeclarations[0].correct[0].target = "GAP_INVALID"
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject when a correct.source is not in gapTexts array", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.responseDeclarations[0].correct[0].source = "TOKEN_INVALID"
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject gap placeholder in top-level body", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.body.unshift({
			type: "paragraph",
			content: [{ type: "gap", gapId: "GAP_1" }]
		})
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject when correct source usage exceeds matchMax", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.interactions.gap_match.gapTexts[0].matchMax = 1
		invalidItem.interactions.gap_match.gaps.push({
			identifier: "GAP_2",
			required: true
		})
		invalidItem.interactions.gap_match.content[0].content.push({
			type: "text",
			content: " and "
		})
		invalidItem.interactions.gap_match.content[0].content.push({
			type: "gap",
			gapId: "GAP_2"
		})
		invalidItem.responseDeclarations[0].correct = [
			{ source: "TOKEN_A", target: "GAP_1" },
			{ source: "TOKEN_A", target: "GAP_2" }
		]
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject duplicate correct pairs", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.responseDeclarations[0].correct = [
			{ source: "TOKEN_A", target: "GAP_1" },
			{ source: "TOKEN_A", target: "GAP_1" }
		]
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should reject ordered cardinality for gapMatch", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.responseDeclarations[0].cardinality = "ordered"
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})

	test("should compile a valid gapMatchInteraction successfully", async () => {
		const xml = await compile(baseValidItem, allWidgetsCollection)
		expect(xml).toContain("<qti-gap-match-interaction")
		expect(xml).toContain(
			'<qti-gap identifier="GAP_1" style="display: inline-block; min-width: 320px; min-height: 120px; border: 2px dashed #ccc; padding: 12px; vertical-align: middle;"/>'
		)
		expect((xml.match(/<qti-gap identifier=/g) || []).length).toBe(1)
		expect(xml).toMatchSnapshot()
	})

	test("should allow matchMax=0 (unlimited) with multiple usages", async () => {
		const validItem = JSON.parse(JSON.stringify(baseValidItem))
		validItem.interactions.gap_match.gapTexts[0].matchMax = 0 // Unlimited
		validItem.interactions.gap_match.gaps.push({
			identifier: "GAP_2",
			required: true
		})
		validItem.interactions.gap_match.content[0].content.push({
			type: "text",
			content: " and "
		})
		validItem.interactions.gap_match.content[0].content.push({
			type: "gap",
			gapId: "GAP_2"
		})
		validItem.responseDeclarations[0].correct = [
			{ source: "TOKEN_A", target: "GAP_1" },
			{ source: "TOKEN_A", target: "GAP_2" }
		]
		const xml = await compile(validItem, allWidgetsCollection)
		expect(xml).toContain("<qti-gap-match-interaction")
		expect((xml.match(/<qti-gap identifier=/g) || []).length).toBe(2)
	})

	test("should reject gap placeholder in feedback blocks", async () => {
		const invalidItem = JSON.parse(JSON.stringify(baseValidItem))
		invalidItem.feedback.FEEDBACK__OVERALL.RESPONSE.CORRECT.content = [
			{
				type: "paragraph",
				content: [{ type: "gap", gapId: "GAP_1" }]
			}
		]
		await expect(compile(invalidItem, allWidgetsCollection)).rejects.toThrow()
	})
})
