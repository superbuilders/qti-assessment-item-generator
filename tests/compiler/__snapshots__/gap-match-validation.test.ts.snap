// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Compiler: gapMatchInteraction validation should compile a valid gapMatchInteraction successfully 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<qti-assessment-item
    xmlns="http://www.imsglobal.org/xsd/imsqtiasi_v3p0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.imsglobal.org/xsd/imsqtiasi_v3p0 https://purl.imsglobal.org/spec/qti/v3p0/schema/xsd/imsqti_asiv3p0p1_v1p0.xsd http://www.w3.org/1998/Math/MathML https://purl.imsglobal.org/spec/mathml/v3p0/schema/xsd/mathml3.xsd"
    identifier="gap-match-validation-item"
    title="Gap Match Validation"
    time-dependent="false"
    xml:lang="en-US">
    <qti-response-declaration identifier="RESPONSE" cardinality="multiple" base-type="directedPair">
        <qti-correct-response>
            <qti-value>TOKEN_A GAP_1</qti-value>
        </qti-correct-response>
        <qti-mapping lower-bound="0" upper-bound="1" default-value="0">
            <qti-map-entry map-key="TOKEN_A GAP_1" mapped-value="1"/>
        </qti-mapping>
    </qti-response-declaration>
    <qti-outcome-declaration identifier="SCORE" cardinality="single" base-type="float">
        <qti-default-value><qti-value>0</qti-value></qti-default-value>
    </qti-outcome-declaration>
    <qti-outcome-declaration identifier="FEEDBACK__OVERALL" cardinality="single" base-type="identifier"/>
    <qti-item-body>
        <qti-gap-match-interaction response-identifier="RESPONSE" shuffle="true" max-associations="1">
            <qti-gap-text identifier="TOKEN_A" match-max="1">A</qti-gap-text>
            <p>Fill blank: <qti-gap identifier="GAP_1" style="display: inline-block; min-width: 320px; min-height: 120px; border: 2px dashed #ccc; padding: 12px; vertical-align: middle;"/></p>
        
        </qti-gap-match-interaction>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_CORRECT" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="correct" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#047857;">Correct! Fantastic work.</p>
      <p style="margin:8px 0 0 0; color:#111827;">Your answer is correct.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Confirmation</p>
        <p>You have demonstrated understanding.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Next Steps</p>
        <p>Continue to the next question.</p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_INCORRECT" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="incorrect" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#B91C1C;">Not quite! Try again.</p>
      <p style="margin:8px 0 0 0; color:#111827;">Your answer is incorrect.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Identify Misconception</p>
        <p>Let's review the concept.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Worked Example</p>
        <p>Here is how to solve it.</p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
    </qti-item-body>
    <qti-response-processing>
        
    <qti-response-condition>
        <qti-response-if>
            <qti-match>
                <qti-variable identifier="RESPONSE"/>
                <qti-correct identifier="RESPONSE"/>
            </qti-match>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_CORRECT</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-if>
        <qti-response-else>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_INCORRECT</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-else>
    </qti-response-condition>
    <qti-response-condition>
        <qti-response-if>
            <qti-and>
                <qti-match><qti-variable identifier="RESPONSE"/><qti-correct identifier="RESPONSE"/></qti-match>
            </qti-and>
            <qti-set-outcome-value identifier="SCORE"><qti-base-value base-type="float">1.0</qti-base-value></qti-set-outcome-value>
        </qti-response-if>
        <qti-response-else>
            <qti-set-outcome-value identifier="SCORE"><qti-base-value base-type="float">0.0</qti-base-value></qti-set-outcome-value>
        </qti-response-else>
    </qti-response-condition>
    </qti-response-processing>
</qti-assessment-item>"
`;
